### PR TITLE
Preserve state of wazuh-indexer service on upgrade

### LINKED ISSUE #487: Preserve previous state of `wazuh-indexer.service` on upgrade
## Description

We are aiming for a consistent behavior of our packages during the lifecycle. Unlike the `wazuh-agent` and `wazuh-manager` services, our service does not preserve the previous state of the service when performing an upgrade. The service must be enabled and restarted manually afterward.

The `wazuh-indexer` post installation scripts must be updated to preserve the previous service status, as other Wazuh central components packages do.

**Expected behavior**

| Before update                   | After update                    |
| ------------------------------- | ------------------------------- |
| `wazuh-indexer.service` stopped | `wazuh-indexer.service` stopped |
| `wazuh-indexer.service` running | `wazuh-indexer.service` running |

### PR DESCRIPTION
**Pull request by QU3B1M**
### Description

Update wazuh-indexer pre, and post installation scripts to preserve the previous service status, maintaining consistency with the rest of Wazuh central components packages.

### Related Issues
Resolves #487

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

---

### Testing evidence

**Upgrade .deb package validation**

1. <details><summary>Upgrade with service running</summary>

    - Service is running before upgrade
        ```bash
        systemctl status wazuh-indexer
        ● wazuh-indexer.service - wazuh-indexer
             Loaded: loaded (/lib/systemd/system/wazuh-indexer.service; enabled; preset: enabled)
             Active: active (running) since Fri 2024-10-25 07:22:28 PDT; 7s ago
               Docs: https://documentation.wazuh.com
           Main PID: 4018 (java)
              Tasks: 72 (limit: 4571)
             Memory: 1.4G
                CPU: 14.054s
             CGroup: /system.slice/wazuh-indexer.service
                     └─4018 /usr/share/wazuh-indexer/jdk/bin/java -Xshare:auto -Dopensearch.networkaddress.cache.ttl=60 -Dope>
        ```
    - Upgrade wazuh-indexer
        ```bash
        dpkg -i wazuh-indexer_5.0.0-0_arm64_b2d1265.deb 
        (Reading database ... 151060 files and directories currently installed.)
        Preparing to unpack wazuh-indexer_5.0.0-0_arm64_da70239.deb ...
        Running Wazuh Indexer Pre-Installation Script
        Stop existing wazuh-indexer.service
        Unpacking wazuh-indexer (5.0.0-0) over (5.0.0-0) ...
        Setting up wazuh-indexer (5.0.0-0) ...
        Running Wazuh Indexer Post-Installation Script
        Restarting wazuh-indexer service...
        ```
    - Service is running after upgrade
        ```bash
        systemctl status wazuh-indexer
        ● wazuh-indexer.service - wazuh-indexer
             Loaded: loaded (/lib/systemd/system/wazuh-indexer.service; enabled; preset: enabled)
             Active: active (running) since Fri 2024-10-25 07:23:23 PDT; 7s ago
               Docs: https://documentation.wazuh.com
           Main PID: 4313 (java)
              Tasks: 73 (limit: 4571)
             Memory: 1.3G
                CPU: 13.464s
             CGroup: /system.slice/wazuh-indexer.service
                     └─4313 /usr/share/wazuh-indexer/jdk/bin/java -Xshare:auto -Dopensearch.networkaddress.cache.ttl=60 -Dope>
        ```
    </details>

2. <details><summary>Upgrade with service stopped</summary>

    - Service is stopped before upgrade
        ```bash
        systemctl status wazuh-indexer
        ○ wazuh-indexer.service - wazuh-indexer
             Loaded: loaded (/lib/systemd/system/wazuh-indexer.service; enabled; preset: enabled)
             Active: inactive (dead) since Fri 2024-10-25 07:25:05 PDT; 8s ago
           Duration: 1min 41.067s
               Docs: https://documentation.wazuh.com
            Process: 4313 ExecStart=/usr/share/wazuh-indexer/bin/systemd-entrypoint -p ${PID_DIR}/wazuh-indexer.pid --quiet (>
           Main PID: 4313 (code=exited, status=143)
                CPU: 16.562s
        ```
    - Upgrade wazuh-indexer
        ```
        (Reading database ... 151060 files and directories currently installed.)
        Preparing to unpack wazuh-indexer_5.0.0-0_arm64_da70239.deb ...
        Running Wazuh Indexer Pre-Installation Script
        Unpacking wazuh-indexer (5.0.0-0) over (5.0.0-0) ...
        Setting up wazuh-indexer (5.0.0-0) ...
        Running Wazuh Indexer Post-Installation Script
        ### NOT starting on installation, please execute the following statements to configure wazuh-indexer service to start automatically using systemd
         sudo systemctl daemon-reload
         sudo systemctl enable wazuh-indexer.service
        ### You can start wazuh-indexer service by executing
         sudo systemctl start wazuh-indexer.service
        ```
    - Service is stopped after upgrade
        ```bash
        systemctl status wazuh-indexer
        ○ wazuh-indexer.service - wazuh-indexer
             Loaded: loaded (/lib/systemd/system/wazuh-indexer.service; enabled; preset: enabled)
             Active: inactive (dead) since Fri 2024-10-25 07:25:05 PDT; 19s ago
           Duration: 1min 41.067s
               Docs: https://documentation.wazuh.com
           Main PID: 4313 (code=exited, status=143)
                CPU: 16.562s
        ```

</details>

**Upgrade .rpm package validation**

1. <details><summary>Upgrade with service running</summary>

    - Service is running before upgrade
        ```bash
        systemctl status wazuh-indexer
        ● wazuh-indexer.service - wazuh-indexer
             Loaded: loaded (/usr/lib/systemd/system/wazuh-indexer.service; disabled; preset: disabled)
             Active: active (running) since Fri 2024-10-25 14:14:44 -03; 1min 4s ago
               Docs: https://documentation.wazuh.com
           Main PID: 38161 (java)
              Tasks: 69 (limit: 22569)
             Memory: 1.3G
                CPU: 15.619s
             CGroup: /system.slice/wazuh-indexer.service
                     └─38161 /usr/share/wazuh-indexer/jdk/bin/java -Xshare:auto -Dopensearch.networkaddress.cache.ttl=60 -Dop>

        ```
    - Upgrade wazuh-indexer
        ```bash
        rpm -i wazuh-indexer_5.0.0-0_aarch64_da70239.rpm --force 
        Stop existing wazuh-indexer.service
        Restarting wazuh-indexer service...

        ```
    - Service is running after upgrade
        ```bash
        ● wazuh-indexer.service - wazuh-indexer
             Loaded: loaded (/usr/lib/systemd/system/wazuh-indexer.service; disabled; preset: disabled)
             Active: active (running) since Fri 2024-10-25 14:31:53 -03; 14s ago
               Docs: https://documentation.wazuh.com
           Main PID: 38490 (java)
              Tasks: 72 (limit: 22569)
             Memory: 1.4G
                CPU: 13.157s
             CGroup: /system.slice/wazuh-indexer.service
                     └─38490 /usr/share/wazuh-indexer/jdk/bin/java -Xshare:auto -Dopensearch.networkaddress.cache.ttl=60 -Dop>


        ```
    </details>

2. <details><summary>Upgrade with service stopped</summary>

    - Service is stopped before upgrade
        ```bash
        systemctl status wazuh-indexer
        ○ wazuh-indexer.service - wazuh-indexer
             Loaded: loaded (/usr/lib/systemd/system/wazuh-indexer.service; disabled; preset: disabled)
             Active: inactive (dead)
               Docs: https://documentation.wazuh.com
        ```
    - Upgrade wazuh-indexer
        ```
        rpm -i wazuh-indexer_5.0.0-0_aarch64_da70239.rpm --force 
        ### NOT starting on installation, please execute the following statements to configure wazuh-indexer service to start automatically using systemd
         sudo systemctl daemon-reload
         sudo systemctl enable wazuh-indexer.service
        ### You can start wazuh-indexer service by executing
         sudo systemctl start wazuh-indexer.service
        ```
    - Service is stopped after upgrade
        ```bash
        systemctl status wazuh-indexer
        ○ wazuh-indexer.service - wazuh-indexer
             Loaded: loaded (/usr/lib/systemd/system/wazuh-indexer.service; disabled; preset: disabled)
             Active: inactive (dead)
               Docs: https://documentation.wazuh.com
        ```

</details>

### CODE REVIEW

**Review by mcasas993 — APPROVED** [review]

File: distribution/packages/src/rpm/wazuh-indexer.rpm.spec [hunk 1]
```diff
@@ -162,6 +162,7 @@ set -e
 if command -v systemctl >/dev/null && systemctl is-active %{name}.service >/dev/null; then
     echo "Stop existing %{name}.service"
     systemctl --no-reload stop %{name}.service
+    touch %{tmp_dir}/wazuh-indexer.restart
```
Comment (AlexRuiz7) [inline distribution/packages/src/rpm/wazuh-indexer.rpm.spec:165]: Consider using [systemd-tmpfiles](https://www.baeldung.com/linux/systemd-tmpfiles-configure-temporary-files)

File: distribution/packages/src/rpm/wazuh-indexer.rpm.spec [hunk 2]
```diff
@@ -204,6 +205,15 @@ if command -v systemd-tmpfiles > /dev/null; then
     systemd-tmpfiles --create %{name}.conf
 fi
 
+if [ -f %{tmp_dir}/wazuh-indexer.restart ]; then
+    rm -f %{tmp_dir}/wazuh-indexer.restart
```
Comment (AlexRuiz7) [inline distribution/packages/src/rpm/wazuh-indexer.rpm.spec:209]: Consider using [systemd-tmpfiles](https://www.baeldung.com/linux/systemd-tmpfiles-configure-temporary-files)

File: distribution/packages/src/deb/debian/preinst [hunk 3]
```diff
@@ -13,10 +13,14 @@ set -e
...
 echo "Running Wazuh Indexer Pre-Installation Script"
 
+# Reference to restore actual service status
+restart_service=/tmp/wazuh-indexer.restart
+
 # Stop existing service
 if command -v systemctl >/dev/null && systemctl is-active wazuh-indexer.service >/dev/null; then
     echo "Stop existing wazuh-indexer.service"
     systemctl --no-reload stop wazuh-indexer.service
+    touch $restart_service
```
Comment (AlexRuiz7) [inline distribution/packages/src/deb/debian/preinst:23]: Consider using [systemd-tmpfiles](https://www.baeldung.com/linux/systemd-tmpfiles-configure-temporary-files)

**Comment by QU3B1M** [comment]
### Update regarding [systemd-tmpfiles](https://www.baeldung.com/linux/systemd-tmpfiles-configure-temporary-files):
The tool is already used in our distribution scripts, anyway the implementation for our _.restart temp_ is not that straightforward.

Currently `systemd-tmpfiles` is used in the `postinst` stage for every scenario (`install`|`upgrade`) which is inconvenient since we need to create the `.restart` file only in the cases where the service is _running_ before the upgrade, detecting and defining it in the `preinst` step.

**Possible implementation:**
We could update the `preinst` script to declare into the `wazuh-indexer.conf` the creation of the `.restart` file on the required cases with something like 
```bash
if command -v systemctl >/dev/null && systemctl is-active wazuh-indexer.service >/dev/null; then
    echo "Stop existing wazuh-indexer.service"
    systemctl --no-reload stop wazuh-indexer.service
    echo "f    /tmp/wazuh-indexer.restart   0750 wazuh-indexer wazuh-indexer - -" >> /usr/lib/sysctl.d/wazuh-indexer.conf
```
Anyway we must keep the manual removal of the file after the service restart, contemplating the scenario where the user does not power off the system between upgrades.
  > By manual removal I refer to this if statement
  > ```bash
  > if [ -f %{tmp_dir}/wazuh-indexer.restart ]; then
  >     rm -f %{tmp_dir}/wazuh-indexer.restart
  > ```

### Conclusion

The implementation of the tool won't help us simplifying the code nor making it clearer for the reader.

File: distribution/packages/src/deb/debian/preinst (hunk 3 above)
Comment (QU3B1M) [reply distribution/packages/src/deb/debian/preinst:23]: Analysis in the comment https://github.com/wazuh/wazuh-indexer/pull/492#issuecomment-2441861588

File: distribution/packages/src/rpm/wazuh-indexer.rpm.spec (hunk 2 above)
Comment (QU3B1M) [reply distribution/packages/src/rpm/wazuh-indexer.rpm.spec:209]: Analysis in the comment https://github.com/wazuh/wazuh-indexer/pull/492#issuecomment-2441861588

File: distribution/packages/src/rpm/wazuh-indexer.rpm.spec (hunk 1 above)
Comment (QU3B1M) [reply distribution/packages/src/rpm/wazuh-indexer.rpm.spec:165]: Analysis in the comment https://github.com/wazuh/wazuh-indexer/pull/492#issuecomment-2441861588

**Review by AlexRuiz7 — CHANGES_REQUESTED** [review]
Needs to be ported to 4.10.0.

**Comment by QU3B1M** [comment]
Ported to > https://github.com/wazuh/wazuh-indexer/pull/498

**Pull request closed without merging**